### PR TITLE
Vision + Pose Estimation

### PR DESCRIPTION
**Pull request by rakrakon**

### CODE REVIEW

File: src/main/kotlin/frc/robot/lib/PoseEstimation.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        fun averageAmbiguity(ambiguities: List<Double>): Double {
+            return 1.0 / ambiguities.sumOf { 1.0 / it }
+        }
+    }
+
+    fun processVisionMeasurements(multiplier: Double) {
+        val results = vision.results
+
+        for (result in results) {
+            val ambiguities = result.distanceToTargets.map { d -> d * d }
```
Comment (Emma03L) [inline src/main/kotlin/frc/robot/lib/PoseEstimation.kt:50]: I think using the tag area for ambiguity will be better.

File: src/main/kotlin/frc/robot/subsystems/swerve/SwerveDrive.kt [hunk 2]
```diff
@@ -92,14 +92,15 @@ class SwerveDrive private constructor
     }
 
     companion object {
-        @JvmField
-        val odometryLock = ReentrantLock()
-
```
Comment (Emma03L) [inline src/main/kotlin/frc/robot/subsystems/swerve/SwerveDrive.kt:97]: This shouldn't be deleted, and shouldn't be in this PR.

**Review by Emma03L — CHANGES_REQUESTED** [review]

File: src/main/kotlin/frc/robot/lib/PoseEstimation.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                result.timestamp,
+                VecBuilder.fill(
+                    stddev,
+                    stddev,
+                    stddev * SwerveConstants.MAX_OMEGA_VELOCITY / SwerveConstants.MAX_X_Y_VELOCITY
+                )
+            )
+        }
+    }
+}
```
Comment (Emma03L) [inline src/main/kotlin/frc/robot/lib/PoseEstimation.kt:64]: The class should have a function to get the estimated pose and the distance to the speaker.

File: src/main/kotlin/frc/robot/lib/PoseEstimation.kt (hunk 1 above)
Comment (rakrakon) [reply src/main/kotlin/frc/robot/lib/PoseEstimation.kt:50]: We will see tommorow...

File: src/main/kotlin/frc/robot/lib/PoseEstimation.kt (hunk 3 above)
Comment (rakrakon) [reply src/main/kotlin/frc/robot/lib/PoseEstimation.kt:75]: 1c9885377fe0adeec8b43a787020d4b40d7d8d8f
c45032c092ded7705a2f7f010c332838f0166a60

Comment (rakrakon) [reply src/main/kotlin/frc/robot/lib/PoseEstimation.kt:75]: Function to get the estimated pose should be in the swerve. Not relevant to this Pull Request.

File: src/main/kotlin/frc/robot/subsystems/vision/PhotonVisionIOSim.kt [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        pose.x, pose.y, 0.0, Rotation3d(0.0, 0.0, pose.rotation.radians)
+    )
+
+    override fun updateInputs() {
+        val botPose = SwerveDrive.getInstance().estimator.estimatedPosition
+        val botPose3d = pose2dToPose3d(botPose)
+        val latestResult: PhotonPipelineResult =
+            simCamera.process(
+                0.0,
+                botPose3d.plus(robotToCam.div(-1.0)),
```
Comment (katzuv) [inline src/main/kotlin/frc/robot/subsystems/vision/PhotonVisionIOSim.kt:42]:
```suggestion
                botPose3d + robotToCam / -1.0,
```
But wouldn't that work?
```suggestion
                botPose3d + robotToCam.inverse(),
```

File: src/main/kotlin/frc/robot/subsystems/vision/PhotonVisionIOSim.kt [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        VisionResult(inputs.poseFieldOriented, inputs.timestamp, inputs.distanceToTargets, inputs.poseAmbiguities)
+
+    private fun pose2dToPose3d(pose: Pose2d): Pose3d = Pose3d(
+        pose.x, pose.y, 0.0, Rotation3d(0.0, 0.0, pose.rotation.radians)
+    )
+
+    override fun updateInputs() {
+        val botPose = SwerveDrive.getInstance().estimator.estimatedPosition
+        val botPose3d = pose2dToPose3d(botPose)
+        val latestResult: PhotonPipelineResult =
```
Comment (katzuv) [inline src/main/kotlin/frc/robot/subsystems/vision/PhotonVisionIOSim.kt:39]:
```suggestion
        val latestResult =
```

File: src/main/kotlin/frc/robot/subsystems/vision/Vision.kt [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+            )
+        }
+    }
+
+    override fun periodic() {
+        results.clear()
+        for (io: VisionIO in ios) {
+            val latestResult = io.getLatestResult()
+            results.add(latestResult)
+        }
```
Comment (katzuv) [inline src/main/kotlin/frc/robot/subsystems/vision/Vision.kt:33]:
```suggestion
        for (io in ios) {
            results.add(io.getLatestResult())
        }
```

**Pull request closed without merging**